### ISSUE TITLE
AINodeStorage.cpp:324:28: error: possibly dangling reference to a temporary

### ISSUE THREAD

**Issue by Alexander-Wilms** [vcmi/vcmi#2866]
**Describe the bug**
When I try to build the latest commit (f50f32fade02432d77e8def412f9b313718e3587) using the linux-gcc-test CMake preset, I encounter this error:

```
alexander@PC:~/Technik/Development/vcmi> cmake --build --preset linux-gcc-test
[7/653] Building CXX object AI/VCAI/CMakeFiles/VCAI.dir/Pathfinding/AINodeStorage.cpp.o
FAILED: AI/VCAI/CMakeFiles/VCAI.dir/Pathfinding/AINodeStorage.cpp.o 
/usr/bin/g++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CHRONO_DYN_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_LOCALE_DYN_LINK -DBOOST_LOCALE_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DENABLE_EDITOR -DENABLE_LAUNCHER -DENABLE_QT_TRANSLATIONS -DM_BIN_DIR=\"/usr/local/bin\" -DM_DATA_DIR=\"/usr/local/share/vcmi\" -DM_LIB_DIR=\"/usr/local/lib64/vcmi\" -DVCAI_EXPORTS -DVCMI_NO_EXTRA_VERSION -DVCMI_VERSION_MAJOR=1 -DVCMI_VERSION_MINOR=4 -DVCMI_VERSION_PATCH=0 -DVCMI_VERSION_STRING=\"1.4.0\" -I/home/alexander/Technik/Development/vcmi/AI/VCAI -I/home/alexander/Technik/Development/vcmi/lib -I/home/alexander/Technik/Development/vcmi/lib/.. -I/home/alexander/Technik/Development/vcmi/lib/../include -Wall -Wextra -Wpointer-arith -Wuninitialized -Wmismatched-tags -Wno-unused-parameter -Wno-switch -Wno-reorder -Wno-sign-compare -Wno-varargs -Werror -Wno-error=array-bounds -O2 -g -DNDEBUG -std=gnu++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Winvalid-pch -include /home/alexander/Technik/Development/vcmi/out/build/linux-gcc-test/AI/VCAI/CMakeFiles/VCAI.dir/cmake_pch.hxx -MD -MT AI/VCAI/CMakeFiles/VCAI.dir/Pathfinding/AINodeStorage.cpp.o -MF AI/VCAI/CMakeFiles/VCAI.dir/Pathfinding/AINodeStorage.cpp.o.d -o AI/VCAI/CMakeFiles/VCAI.dir/Pathfinding/AINodeStorage.cpp.o -c /home/alexander/Technik/Development/vcmi/AI/VCAI/Pathfinding/AINodeStorage.cpp
/home/alexander/Technik/Development/vcmi/AI/VCAI/Pathfinding/AINodeStorage.cpp: In member function ‘bool AINodeStorage::isTileAccessible(const int3&, EPathfindingLayer) const’:
/home/alexander/Technik/Development/vcmi/AI/VCAI/Pathfinding/AINodeStorage.cpp:324:28: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  324 |         const AIPathNode & node = nodes[layer][pos.z][pos.x][pos.y][0];
      |                            ^~~~
/home/alexander/Technik/Development/vcmi/AI/VCAI/Pathfinding/AINodeStorage.cpp:324:70: note: the temporary was destroyed at the end of the full expression ‘boost::detail::multi_array::const_sub_array<T, NumDims, TPtr>::operator[](index) const [with T = AIPathNode; long unsigned int NumDims = 2; TPtr = const AIPathNode*; const_reference = boost::detail::multi_array::const_sub_array<AIPathNode, 1, const AIPathNode*>; index = long int](((boost::detail::multi_array::const_sub_array<AIPathNode, 2, const AIPathNode*>::index)((si32)pos.int3::y))).boost::detail::multi_array::const_sub_array<AIPathNode, 1, const AIPathNode*>::operator[](0)’
  324 |         const AIPathNode & node = nodes[layer][pos.z][pos.x][pos.y][0];
      |                                                                      ^
cc1plus: all warnings being treated as errors
[16/653] Building CXX object AI/VCAI/CMakeFiles/VCAI.dir/Goals/Explore.cpp.o
ninja: build stopped: subcommand failed.
```

https://github.com/vcmi/vcmi/blob/develop/AI/VCAI/Pathfinding/AINodeStorage.cpp#L324C4-L324C4

**Game logs**
na

**To Reproduce**
1. git clone git@github.com:vcmi/vcmi.git
2. cd vcmi
3. git checkout f50f32fade02432d77e8def412f9b313718e3587
4. cmake --preset linux-gcc-test
5. cmake --build --preset linux-gcc-test

**Expected behavior**
Build succeeds without any changes

**Actual behavior**
The build fails

**Did it work earlier?**
I usually build it like this, which works fine:

```
mkdir build
cd build
cmake ..
make -j4
```

Since this doesn't use the linux-gcc-test preset and thus not the following [flag](https://github.com/vcmi/vcmi/blob/f50f32fade02432d77e8def412f9b313718e3587/CMakePresets.json#L27) `"ENABLE_STRICT_COMPILATION": "ON",` , the build succeeds.

I tried using the cmake preset, since that's how VCMI is built during CI and encountered this issue.

**Screenshots**
na

**Version**
 - OS: openSUSE Tumbleweed 20230915
 - Version: f50f32fade02432d77e8def412f9b313718e3587

**Additional context**
Looks like this doesn't occur during CI since the `dangling-reference` warning is a recent addition to gcc and false positives are possible: https://trofi.github.io/posts/264-gcc-s-new-Wdangling-reference-warning.html

```
gcc --version
gcc (SUSE Linux) 13.2.1 20230912 [revision b96e66fd4ef3e36983969fb8cdd1956f551a074b]
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by nullkiller** [comment]
These links never survive longer than 1 turn. And array may actually be resized only when map changes. Can be safely suppressed

**Comment by Alexander-Wilms** [comment]
So once Github updates its runner images to include ubuntu-23.04, something like this would need to be added to CMakeLists.txt:

```
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR NOT WIN32)
	[...]
	if(ENABLE_STRICT_COMPILATION)
		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=array-bounds") # false positives in boost::multiarray during release build, keep as warning-only
		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=dangling-reference") # false positives in AI/VCAI/Pathfinding/AINodeStorage.cpp
	endif()
```

**Comment by IvanSavenko** [comment]
Why not change that method to 
```
	return nodes[layer][pos.z][pos.x][pos.y][0].action != EPathNodeAction::UNKNOWN;
```

Not sure in which cases boost multiarray returns temporary, but should fix all of them, even though this form is slightly less readable

**Comment by nullkiller** [comment]
The problem that these references are at least stored in nodes themselves as previous. So partially it can be fixed but not fully

**Comment by IvanSavenko** [comment]
That's not what this warning is about though.
Warning is about cases like this one:
```cpp
int & foo()
{
    int result = 42;
    return result;
}

int main()
{
    const int & bar = foo(); // Illegal! bar now points to local and no longer existing 'result' varaible
}
```

So, in this particular case, gcc believes that expression `nodes[layer][pos.z][pos.x][pos.y][0]` return a reference to a variable local to a operator[].

In fact, this might be a bug in gcc: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108165

So personally I would rather change this code to one-liner as described above or add pragma around this messages to silence this warning *specifically* in this call

**Comment by nullkiller** [comment]
This is VCAI. I thought about Nullkiller and started looking there and found that it is all implemented through
STRONG_INLINE
	```
boost::detail::multi_array::sub_array<AIPathNode, 1> get(int3 tile, EPathfindingLayer layer) const
	{
		return (*nodes)[layer][tile.z][tile.x][tile.y];
	}
```

**Comment by nullkiller** [comment]
Anyway sure one-liner is possible but until nodes array is resized these references are all valid. It is not temporary as we use const AIPathNode &
It is direct reference onto underlying array cell

**Comment by nullkiller** [comment]
At least it is what I expect it to be.

**Comment by Kuxe** [comment]
Submitted PR to fix this error: #3399
